src/polyhedralGravity/util/UtilityFloatArithmetic.cpp

```diff
@@ -0,0 +1,51 @@
+#include "UtilityFloatArithmetic.h"
+
+
+namespace polyhedralGravity::util {
+
+    template<typename FloatType>
+    bool almostEqualUlps(FloatType lhs, FloatType rhs, int ulpDistance) {
+        static_assert(std::is_same_v<FloatType, float> || std::is_same_v<FloatType, double>,
+                      "Template argument must be FloatType: Either float or double!");
+
+        // In case the floats are equal in their representation, return true
+        if (lhs == rhs) {
+            return true;
+        }
+
+        // In case the signs mismatch, return false
+        if (lhs < static_cast<FloatType>(0.0) && rhs > static_cast<FloatType>(0.0) ||
+            lhs > static_cast<FloatType>(0.0) && rhs < static_cast<FloatType>(0.0)) {
+            return false;
+            }
+
+        if constexpr (std::is_same_v<FloatType, float>) {
+            // In case of float, compute ULP distance by interpreting float as 32-bit integer
+            return reinterpret_cast<std::int32_t&>(rhs) - reinterpret_cast<std::int32_t&>(lhs) <= ulpDistance;
+        }
+        else if constexpr (std::is_same_v<FloatType, double>) {
+            // In case of double, compute ULP distance by interpreting double as 64-bit integer
+            return reinterpret_cast<std::int64_t&>(rhs) - reinterpret_cast<std::int64_t&>(lhs) <= ulpDistance;
+        }
+
+        // Due to the static_assert above, this should not happen
+        return false;
+    }
+
+    // Template Instantations for float and double (required since definition is in .cpp file,
+    //      also makes the static assert not strictly necessary)
+    template bool almostEqualUlps<float>(float lhs, float rhs, int ulpDistance);
+    template bool almostEqualUlps<double>(double lhs, double rhs, int ulpDistance);
+
+
+    template<typename FloatType>
+    bool almostEqualRelative(FloatType lhs, FloatType rhs, double epsilon) {
+        const FloatType diff = std::abs(rhs - lhs);
+        const FloatType largerValue = std::max(std::abs(rhs), std::abs(lhs));
+        return diff <= largerValue * epsilon;
+    }
```

Comment:
I think this makes the assumption that the larger value is the ground truth, which is not necessarily the case.

E.g. 100 and 25. If 25 is truth, then relative error is 3. Or 0.75 vice versa.

In that case 
```
diff =  75 ,
larger = 100
```

so you compare
```
75 <= 100 * eps 
```

which would yield true for eps >= 0.75, so only correct if 100 is ground truth.

Reply:
I'm not sure if I get the point fully:
An `epsilon >= 0.75` yields true, but it should return false. Correct?

The solution, which I currently included, was suggested by [Bruce Dawson](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/) redirected from the [second post here](https://stackoverflow.com/questions/17333/how-do-you-compare-float-and-double-while-accounting-for-precision-loss).
At least, the accepted answers all seem to have the mentioned issue (given that I understood the feedback?) with bigger epsilons > 1.0.

Should I add an`if` to check that the epsilon is smaller than e.g. `1.0` like [here](https://stackoverflow.com/questions/4915462/how-should-i-do-floating-point-comparison)?

Reply:
Okay, on second thought I think, it is fine to do this here because you really only care, if the values are similar and they are presumably always very close, it's just not a formally correct relative error terms

<details>
  <summary>What I was thinking about (and uh you can have spoilers in code comments, fancy :D :D)</summary>
See https://en.wikipedia.org/wiki/Approximation_error#Formal_definition

The problem is not eps > 1, the problem is the question of relative to what? Typically, it is relative to the correct value (v in below equation)

![image](https://github.com/esa/polyhedral-gravity-model/assets/2656685/934dad16-fe9d-48a3-bc1d-5f683081904f)

By replacing it with the max(v,v_approx) you have an ambiguous definition of relative (that also happens to be biased towards smaller errors :P . For the functionality here it doesn't matter much ofc, but this is not really a relative error.
</details>

Reply:
Okay. I get it. Thank you for clarifying 👍